editor/src/clj/editor/code/view.clj

```diff
@@ -3030,31 +2972,40 @@
 
 (defn- setup-view! [resource-node view-node app-view lsp]
   ;; Grab the unmodified lines or io error before opening the
-  ;; file. Otherwise this will happen on the first edit. If a
+  ;; file. Otherwise, this will happen on the first edit. If a
   ;; background process has modified (or even deleted) the file
   ;; without the editor knowing, the "original" unmodified lines
   ;; reached after a series of undo's could be something else entirely
   ;; than what the user saw.
-  (g/with-auto-evaluation-context ec
-    (r/ensure-loaded! resource-node ec)
-    (let [glyph-metrics (g/node-value view-node :glyph-metrics ec)
-          tab-spaces (g/node-value view-node :tab-spaces ec)
-          tab-stops (data/tab-stops glyph-metrics tab-spaces)
-          lines (g/node-value resource-node :lines ec)
-          document-width (data/max-line-width glyph-metrics tab-stops lines)
-          resource (g/node-value resource-node :resource ec)]
-      (when (resource/file-resource? resource)
-        (lsp/open-view! lsp view-node resource lines))
+  (let [resource (resource-node/resource resource-node)
+        resource-type (resource/resource-type resource)
+        resource-node-type (:node-type resource-type)
+        is-code-resource-type (r/code-resource-type? resource-type)]
+    (when is-code-resource-type
+      (g/with-auto-evaluation-context evaluation-context
+        (r/ensure-loaded! resource-node evaluation-context)))
+    (let [[lines document-width]
+          (g/with-auto-evaluation-context evaluation-context
+            (let [glyph-metrics (g/node-value view-node :glyph-metrics evaluation-context)
+                  tab-spaces (g/node-value view-node :tab-spaces evaluation-context)
+                  tab-stops (data/tab-stops glyph-metrics tab-spaces)
+                  lines (g/node-value resource-node :lines evaluation-context)
+                  document-width (data/max-line-width glyph-metrics tab-stops lines)]
+              [lines document-width]))]
```

Comment:
This was slightly wrong before. Technically you should not use an `evaluation-context` after you've made any modifications to the graph, because the changes will not be seen from within the `evaluation-context`.

I've refactored the code a bit so that we make a new `evaluation-context` after we've loaded the file, but the main change here is that we only call `code.resource/ensure-loaded!` to force lazy-loaded `CodeEditorResourceNodes`. We don't want to do it for other node types.

Reply:
how critical it is? is it only about text metrics reading, or properties as well?

Reply:
Basically, if a resource type that has been registered using the `register-code-resource-type` function specifies `:lazy-loaded true`, its `:load-fn` will not be executed at all when the project is loaded. So it won't have any content at all. You can't even show the lines of the file without loading them. Calling `ensure-loaded!` on it makes sure that it is fully loaded into memory if it hasn't yet been loaded for some other reason.

Lazy-loading is just a thing for `CodeEditorResourceNode`-derived node types, which covers all the code-editable resource types such as scripts, shaders, as well as game-specific JSON, XML, and so on.

We introduced lazy-loading because game projects often have quite a few very large JSON or XML files that is not relevant to the Defold editor, but are used in the game at runtime. For example, many projects have a few massive JSON or XML files that contain localized strings and so on.